src/carp.cc

```diff
@@ -207,16 +208,11 @@ carpSelectParent(PeerSelector *ps)
         debugs(39, 3, "carpSelectParent: key=" << key << " name=" << tp->name << " combined_hash=" << combined_hash  <<
                " score=" << std::setprecision(0) << score);
 
-        if ((score > high_score) && peerHTTPOkay(tp, ps)) {
-            p = tp;
-            high_score = score;
-        }
+        if (peerHTTPOkay(tp, ps))
+            sortedPeers.push_back(std::make_pair(score, tp));
     }
 
-    if (p)
-        debugs(39, 2, "carpSelectParent: selected " << p->name);
-
-    return p;
+    ps->addGroup(sortedPeers, CARP);
```

Comment:
Thank you to find this bug all this arithmetic confuses if you are looking a lot of time the same part of code.

Actually we need:
 - For round robin we need to select the less accessed peer (modified by a weight)
-  For weighted round robin the same
-  For the neighbours digest the one with the smallest round trip time.
-  For carp we are computing a score and we should select the one with higher score
-  For source_hash and user_hash the one with higher score.

The addGroup must sorting in ascending order, and we should use negative numbers/scores as keys for emulate descending order. The addGroup was not the only one which had bug but the code in carp was buggy too, because requires a list in descending order (a peer with the highest score should selected).

In the latest patch (to be committed):
  - AddGroup adds always in ascending order.
  - Fixed the carp to negate the score when use it with addGroup, to emulate desceding order.

Reply:
The code is enough for someone to understand the functions are small, and these comments does not add something. Also the above descriptions are not accurate.
There is some description here: 
      https://wiki.squid-cache.org/Features/LoadBalance